### PR TITLE
Add S3StorageOptions to allow configuring S3 backend explicitly

### PR DESCRIPTION
**Pull request by xianwill**
# Description

Add S3StorageOptions to allow configuring S3 backend explicitly (rather than through env).

This enables different backend configuration for two different delta tables (which is a need in kafka-delta-ingest for dead lettering to delta). With this change, we can configure two different storage backends with different parameters (specifically DynamoDb lock name difference is needed for our use case) and use the `DeltaTable::new` factory (https://github.com/delta-io/delta-rs/blob/main/rust/src/delta.rs#L1038).

# Related Issue(s)

Does not reference an existing delta-rs issue, but required for dead lettering to delta in kafka-delta-ingest - https://github.com/delta-io/kafka-delta-ingest/pull/69

### CODE REVIEW

File: rust/src/storage/s3/dynamodb_lock.rs
Comment (xianwill) [inline rust/src/storage/s3/dynamodb_lock.rs:69]:
```suggestion
    /// Keys not present in the map are taken from the environment variable.
```

File: rust/src/storage/s3/mod.rs
Comment (rtyler) [inline rust/src/storage/s3/mod.rs:27]: Why are these also not considered options for use with the S3StorageOptions? :confused:  (Particularly endpoint url since that's got some behavioral meaning later on)

File: rust/src/storage/s3/mod.rs
Comment (rtyler) [inline rust/src/storage/s3/mod.rs:85]: what's with the functions in your functions? As opposed to non public functions within the `impl`?

File: rust/src/storage/s3/mod.rs
Comment (rtyler) [inline rust/src/storage/s3/mod.rs:78]: I might recommending moving this to a `Default` trait implementation.

File: rust/src/storage/s3/mod.rs
Comment (xianwill) [reply rust/src/storage/s3/mod.rs:27]: i actually started w endpoint url as an option but that var is only set when testing against localstack and only used to create the region. given that, i felt like it was awkward to include that as part of the public api. we already set the env var in ci so i feel like omitting it from the options struct is a cleaner public api for consumers. i dont think anybody consuming delta-rs should need to use it

File: rust/src/storage/s3/mod.rs
Comment (xianwill) [reply rust/src/storage/s3/mod.rs]: :xzibit:

hoisted these up.

File: rust/src/storage/s3/mod.rs
Comment (xianwill) [reply rust/src/storage/s3/mod.rs]: :heavy_check_mark:

File: rust/src/storage/s3/mod.rs
Comment (xianwill) [reply rust/src/storage/s3/mod.rs]: RE: `AWS_ENDPOINT_URL` - I moved this to S3StorageOptions in the latest commit in case there are other use cases for custom S3 endpoints. The URL is only used for creating the `region`, but I added an unused `_endpoint_url` field for information/debug needs.

RE: `AWS_WEB_IDENTITY_TOKEN_FILE` - these must be set in the env for the rusoto web identity provider to work the way it is called today (`from_k8s_env`). I went ahead added these to the `s3_storage_options` keys along w/ some code to set them in env if they are passed in options map.

Comment (rtyler) [reply rust/src/storage/s3/mod.rs]: `AWS_ENDPOINT_URL` will be required for anybody working with an S3 compatible API but not using S3 directly, which is why I think it should be included. Yes we're using it for localstack, but excluding it from the options limits the potential for users to use non-S3 storage providers with this code :shrug:

Comment (xianwill) [reply rust/src/storage/s3/mod.rs]: Cool. Its in the options now so I think we are good.

**Review by rtyler — APPROVED** [review]
This looks fine to me but a committer with some smarts should actually look at it before merging :)

**Review by mosyp — APPROVED** [review]
LGTM, love it!

**Review by fvaleye — APPROVED** [review]
Nice! 👍

**Pull request merged**
